File: python/infinicore/nn/functional/gaussian_nll_loss.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor

_REDUCTION_TO_INT = {
    "none": 0,
    "mean": 1,
    "sum": 2,
}


def gaussian_nll_loss(
    input: Tensor,
    target: Tensor,
    var: Tensor,
    full: bool = False,
    eps: float = 1e-6,
    reduction: str = "mean",
) -> Tensor:
    reduction_i = _REDUCTION_TO_INT.get(reduction)
    if reduction_i is None:
        raise ValueError(f"Unsupported reduction: {reduction!r}")

    return Tensor(
        _infinicore.gaussian_nll_loss(
            input._underlying,
            target._underlying,
            var._underlying,
            bool(full),
            float(eps),
            int(reduction_i),
        )
    )

File: ops/src/infinicore/ops/gaussian_nll_loss/gaussian_nll_loss.cc
#include "infinicore/ops/gaussian_nll_loss.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(GaussianNllLoss);

GaussianNllLoss::GaussianNllLoss(Tensor out,
                                 const Tensor &input,
                                 const Tensor &target,
                                 const Tensor &var,
                                 bool full,
                                 double eps,
                                 int reduction) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(out, input, target, var);
    INFINICORE_GRAPH_OP_DISPATCH(out->device().getType(), out, input, target, var, full, eps, reduction);
}

void GaussianNllLoss::execute(Tensor out,
                              const Tensor &input,
                              const Tensor &target,
                              const Tensor &var,
                              bool full,
                              double eps,
                              int reduction) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(GaussianNllLoss, out, input, target, var, full, eps, reduction);
}

Tensor gaussian_nll_loss(const Tensor &input,
                         const Tensor &target,
                         const Tensor &var,
                         bool full,
                         double eps,
                         int reduction) {
    std::vector<size_t> out_shape = (reduction == 0) ? input->shape() : std::vector<size_t>{};
    auto out = Tensor::empty(out_shape, input->dtype(), input->device());
    gaussian_nll_loss_(out, input, target, var, full, eps, reduction);
    return out;
}

void gaussian_nll_loss_(Tensor out,
                        const Tensor &input,
                        const Tensor &target,
                        const Tensor &var,
                        bool full,
                        double eps,
                        int reduction) {
    GaussianNllLoss::execute(out, input, target, var, full, eps, reduction);
}

} // namespace infinicore::op
